pkg/controller/clusterdeployment/clusterinstalls.go

```diff
@@ -0,0 +1,239 @@
+package clusterdeployment
+
+import (
+	"context"
+	"reflect"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	hivecontractsv1alpha1 "github.com/openshift/hive/apis/hivecontracts/v1alpha1"
+	"github.com/openshift/hive/pkg/constants"
+	hivemetrics "github.com/openshift/hive/pkg/controller/metrics"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+)
+
+func (r *ReconcileClusterDeployment) reconcileExistingClusterInstall(cd *hivev1.ClusterDeployment, logger log.FieldLogger) (reconcile.Result, error) {
+	ref := cd.Spec.ClusterInstallRef
+	gvk := schema.GroupVersionKind{
+		Group:   ref.Group,
+		Version: ref.Version,
+		Kind:    ref.Kind,
+	}
+
+	logger = logger.WithField("clusterinstall", ref.Name).WithField("gvk", gvk)
+	logger.Debug("reconciling existing clusterinstall")
+
+	ci := &hivecontractsv1alpha1.ClusterInstall{}
+	err := controllerutils.GetDuckType(context.TODO(), r.Client,
+		gvk,
+		types.NamespacedName{Namespace: cd.Namespace, Name: ref.Name},
+		ci)
+	if apierrors.IsNotFound(err) {
+		logger.Debug("cluster is not found, so skipping")
+		return reconcile.Result{}, nil
+	}
+	if err != nil {
+		logger.WithError(err).Error("failed to get the cluster install")
+		return reconcile.Result{}, err
+	}
+
+	specModified := false
+	statusModified := false
+	// copy the cluster metadata
+	if met := ci.Spec.ClusterMetadata; met != nil {
```

Comment:
Something I've been unsure about here is if we should do this before install is successful or not. Currently I believe a ClusterProvision gets metadata before launching installer, and that is immediately copied to ClusterDeployment. It looks like your impl is consistent with that, the metadata appears on ClusterDeployment immediately even if install is underway. If it changes, it's copied again.

Long story short, this looks good to me, consistent with what we do today. :)

Reply:
once the provision has completed, i don't like us syncing changed like cluster metadata from clusterinstall . i thin that info should be available before you go Completed = true level.

Reply:
it provides a good place for our controller to go, i have all the info i need now, the cluster install object is leveled that i can keep my info stable.